lib/cretonne/src/dominator_tree.rs

```diff
@@ -68,8 +68,20 @@ impl DominatorTree {
     /// is unreachable.
     ///
     /// An instruction is considered to dominate itself.
-    pub fn dominates(&self, a: Inst, mut b: Inst, layout: &Layout) -> bool {
+    pub fn dominates(&self, a: Inst, b: Inst, layout: &Layout) -> bool {
         let ebb_a = layout.inst_ebb(a).expect("Instruction not in layout.");
+        self.ebb_dominates(ebb_a, b, layout) && layout.cmp(a, b) != Ordering::Greater
+    }
+
+    /// Returns `true` if `ebb_a` dominates `b`.
+    ///
+    /// This means that every control-flow path from the function entry to `b` must go through
+    /// `ebb_a`.
```

Comment:
I think we should be more precise with the language here. What does it mean to "go through" an EBB? In this case it means "go through the `ebb_a` header". An EBB header can dominate instructions that are not dominated by later instructions in the EBB.